### PR TITLE
Introduce custom title bar

### PR DESCRIPTION
**Pull request by onkofonko**
Introduce a configurable custom title bar.
Editable using CSS
Screenshot 2024-08-30 223413
image

### CODE REVIEW

**Comment by SManyac** [comment]
Nice

**Comment by ThaUnknown** [comment]
this is nice, however:
- needs to be tested on windows 10
- needs to be tested on mac
- needs to be tested on linux
- you're using the wrong code styling, use Miru's ESLint config

**Review by ThaUnknown — CHANGES_REQUESTED** [review]
needs further review and styling fixes

**Comment by onkofonko** [comment]
Will test on various systems and correct code styling.

**Comment by onkofonko** [comment]
Unfortunately, I cannot test on MacOS as I don't have a MacOS device available to build or test.
Linux (Ubuntu):
https://github.com/user-attachments/assets/ffeb1c87-6be1-45b1-91f7-ab7d10294b09

Windows 10:

**Comment by onkofonko** [comment]
Unfortunately, I can't test MacOS Build as I don't own a MacOS device, so I can't even build it.

Linux (Ubuntu):

https://github.com/user-attachments/assets/20b4b9e0-a9fe-4149-9c19-dd1b853e7ee7

Windows 10: 

https://github.com/user-attachments/assets/e53c189a-c761-4337-b44f-fb6b571cae0a

File: common/components/Menubar.svelte [hunk 1]
```diff
@@ -13,10 +13,22 @@
 
   const debug = persisted('debug', '', {
     serializer: {
-      parse: e => e,
-      stringify: e => e
+      parse: (e) => e,
+      stringify: (e) => e
```
Comment (ThaUnknown) [inline common/components/Menubar.svelte:17]: there's no reason to change the styling on these

File: common/components/Menubar.svelte [hunk 2]
```diff
@@ -13,10 +13,22 @@
...
+  function minimize () {
+    window.ipcRenderer.send('minimize')
+  }
+
+  function maximize () {
+    window.ipcRenderer.send('maximize')
+  }
+
+  function closeapp () {
+    window.ipcRenderer.send('closeapp')
```
Comment (ThaUnknown) [inline common/components/Menubar.svelte:30]: im fairly sure miru already implements this functionality

File: common/components/Menubar.svelte [hunk 3]
```diff
@@ -51,10 +61,10 @@
     transform: translate(-29.3%) rotate(-45deg);
   }
   .navbar {
-    --navbar-height: 28px !important;
+    --navbar-height: 32px !important;
```
Comment (ThaUnknown) [inline common/components/Menubar.svelte:64]: why 32px?

File: common/components/Menubar.svelte [hunk 4]
```diff
@@ -51,10 +61,10 @@
...
   }
   .z-101 {
-    z-index: 101 !important
+    z-index: 101 !important;
```
Comment (ThaUnknown) [inline common/components/Menubar.svelte:67]: don't restyle code you havent changed

File: common/components/Menubar.svelte [hunk 5]
```diff
@@ -64,29 +74,68 @@
...
-    left: unset !important
+    left: unset !important;
   }
   @media (pointer: none), (pointer: coarse) {
     .navbar {
       display: none !important;
       height: 0;
     }
   }
+  #window-controls {
```
Comment (ThaUnknown) [inline common/components/Menubar.svelte:93]: dont use ids, use classes

File: electron/src/main/app.js [hunk 6]
```diff
@@ -29,19 +29,16 @@ export default class App {
...
     backgroundColor: '#17191c',
     autoHideMenuBar: true,
     webPreferences: {
       webSecurity: false,
       allowRunningInsecureContent: false,
       enableBlinkFeatures: 'FontAccess, AudioVideoTracks',
       backgroundThrottling: false,
-      preload: join(__dirname, '/preload.js')
+      preload: join(__dirname, '/preload.js'),
+      contextIsolation: true,
```
Comment (ThaUnknown) [inline electron/src/main/app.js:40]: this should be in a separate PR, with an explanation, it's out of scope for this PR

File: electron/src/main/app.js [hunk 7]
```diff
@@ -159,3 +156,28 @@ export default class App {
     if (!this.updater.install(forceRunAfter)) app.quit()
   }
 }
+
+ipcMain.on('minimize', (event) => {
+  const focusedWindow = BrowserWindow.getFocusedWindow()
```
Comment (ThaUnknown) [inline electron/src/main/app.js:161]: absolutely not, miru already defines ways to read ipc messages for the main window, this is not the way of doing it

File: electron/src/preload/preload.js [hunk 8]
```diff
@@ -30,3 +30,8 @@ ipcRenderer.once('port', ({ ports }) => {
     }
   })
 })
+
+contextBridge.exposeInMainWorld('ipcRenderer', {
```
Comment (ThaUnknown) [inline electron/src/preload/preload.js:34]: no, you're repeating the code above, miru already exposes this functionality via `IPC`

**Review by ThaUnknown — CHANGES_REQUESTED** [review]
close but not quite, this is really nice tho

**Comment by onkofonko** [comment]
I will address all the comments within the next few days.

File: common/components/Menubar.svelte (hunk 3 above)
Comment (onkofonko) [reply common/components/Menubar.svelte:64]: The standard title bar has a height of 32px.
Source: https://learn.microsoft.com/en-us/windows/apps/design/basics/titlebar-design

**Comment by onkofonko** [comment]
Retested on Windows 11, Windows 10 and Ubuntu

File: electron/src/main/app.js [hunk 9]
```diff
@@ -65,6 +60,11 @@ export default class App {
     this.mainWindow.on('closed', () => this.destroy())
     this.webtorrentWindow.on('closed', () => this.destroy())
     ipcMain.on('close', () => this.destroy())
+    ipcMain.on('minimize', () => BrowserWindow.getFocusedWindow()?.minimize())
```
Comment (ThaUnknown) [inline electron/src/main/app.js:63]: this should be this.mainWindow

File: common/components/Menubar.svelte [hunk 10]
```diff
@@ -66,27 +64,59 @@
...
   }
   @media (pointer: none), (pointer: coarse) {
     .navbar {
       display: none !important;
       height: 0;
     }
   }
+  .window-controls {
+    display: grid;
+    grid-template-columns: repeat(3, 46px);
```
Comment (ThaUnknown) [inline common/components/Menubar.svelte:83]: is there any reason grid is used here? or is it simply easier to work with?

**Review by ThaUnknown — CHANGES_REQUESTED** [review]
i think the minimize and maximize buttons should also have a slightly transparent hover background, just like the close button, this is a really nice change

**Comment by onkofonko** [comment]
> i think the minimize and maximize buttons should also have a slightly transparent hover background, just like the close button, this is a really nice change

Would you like the hover color to be an accent color?

File: common/components/Menubar.svelte (hunk 10 above)
Comment (onkofonko) [reply common/components/Menubar.svelte:83]: It can also be display:flex

**Comment by ThaUnknown** [comment]
i think its best to keep it accurate to windows, so some semi-transparent white

**Comment by onkofonko** [comment]
> i think its best to keep it accurate to windows, so some semi-transparent white

dark background: 
image

light background: 
image

The best thing would be to make the hover effect and icon colors reactive, but I'm not sure how to do it without affecting the close button's hover color

File: common/components/Menubar.svelte [hunk 11]
```diff
@@ -25,14 +25,12 @@
       <img src='./logo_filled.png' class='position-absolute w-50 h-50 m-10 pointer d-md-block d-none p-5' alt='ico' use:click={close} />
     {/if}
   </div>
+  <div class='window-controls'>
+    <button class='button max-button'  on:click={() => IPC.emit('minimize')}><svg class='svg-controls' height='12' role='img' viewBox='0 0 12 12'width='12'><rect fill='currentColor' height='1' width='10' x='1' y='6' /></svg></button>
+    <button class='button restore-button'  on:click={() => IPC.emit('maximize')}><svg class='svg-controls' height='12' role='img' viewBox='0 0 12 12'width='12'><rect fill='none' height='9' stroke='currentColor' width='9' x='1.5' y='1.5' /></svg></button>
+    <button class='button close-button' on:click={() => IPC.emit('close')}><svg class='svg-controls' height='12' role='img' viewBox='0 0 12 12'width='12'><polygon fill='currentColor' fill-rule='evenodd' points='11 1.576 6.583 6 11 10.424 10.424 11 6 6.583 1.576 11 1 10.424 5.417 6 1 1.576 1.576 1 6 5.417 10.424 1' /></svg></button>
+  </div>
   <div class='h-full bg-dark flex-grow-1'>
```
Comment (ThaUnknown) [inline common/components/Menubar.svelte:33]: this div is now empty, is it needed?

File: common/components/Menubar.svelte [hunk 12]
```diff
@@ -66,27 +64,58 @@
...
+    left: unset !important;
   }
   @media (pointer: none), (pointer: coarse) {
     .navbar {
       display: none !important;
       height: 0;
     }
   }
+  .window-controls {
+    display: flex;
```
Comment (ThaUnknown) [inline common/components/Menubar.svelte:82]: i think all of these styles can be instead used as bootstrap's atomic clasess:
```css
d-flex position-absolute top-0 right-0 height-full
```

File: common/components/Menubar.svelte [hunk 13]
```diff
@@ -66,27 +64,58 @@
...
+  .window-controls .button {
+    grid-row: 1 / span 1;
+    display: flex;
+    justify-content: center;
+    align-items: center;
+    width: 46px;
+    height: 32px;
+    background: transparent;
+    border: none;
+    color: #ffffff;
```
Comment (ThaUnknown) [inline common/components/Menubar.svelte:100]: same as above, border-0 color-white and i think the grid row can go?

**Review by ThaUnknown — APPROVED** [review]
lgtm

**Comment by ThaUnknown** [comment]
this is good to go, but i'm currently on vacation, so I cannot merge this until I get back next month

**Comment by onkofonko** [comment]
No problem. Have fun on your vacation.

**Pull request merged**